airbyte-cdk/python/airbyte_cdk/sources/file_based/writers/local_file_client.py

```diff
@@ -0,0 +1,70 @@
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+
+import logging
+import os
+import time
+from io import IOBase
+from typing import Any, Dict
+
+import psutil
+
+AIRBYTE_STAGING_DIRECTORY = os.getenv("AIRBYTE_STAGING_DIRECTORY", "/staging/files")
+DEFAULT_LOCAL_DIRECTORY = "/tmp/airbyte-file-transfer"
+
+
+class LocalFileTransferClient:
+    def __init__(self) -> None:
+        """
+        Initialize the LocalFileTransferClient. It uses a default local directory for file saving.
+        """
+        self._local_directory = AIRBYTE_STAGING_DIRECTORY if os.path.exists(AIRBYTE_STAGING_DIRECTORY) else DEFAULT_LOCAL_DIRECTORY
+
+    def write(self, file_uri: str, fp: IOBase, file_size: int, logger: logging.Logger) -> Dict[str, Any]:
+        """
+        Write the file to a local directory.
+        """
+        # Remove left slashes from source path format to make relative path for writing locally
+        file_relative_path = file_uri.lstrip("/")
+        local_file_path = os.path.join(self._local_directory, file_relative_path)
+
+        # Ensure the local directory exists
+        os.makedirs(os.path.dirname(local_file_path), exist_ok=True)
+
+        # Get the absolute path
+        absolute_file_path = os.path.abspath(local_file_path)
+
+        # Get available disk space
+        disk_usage = psutil.disk_usage("/")
+        available_disk_space = disk_usage.free
+
+        # Get available memory
+        memory_info = psutil.virtual_memory()
+        available_memory = memory_info.available
+
+        # logger.info(f"Writing file to {local_file_path}.")
+        # Log file size, available disk space, and memory
+        logger.info(
+            f"Writing file to '{local_file_path}' "
+            f"with size: {file_size / (1024 * 1024):,.2f} MB ({file_size / (1024 * 1024 * 1024):.2f} GB), "
+            f"available disk space: {available_disk_space / (1024 * 1024):,.2f} MB ({available_disk_space / (1024 * 1024 * 1024):.2f} GB),"
+            f"available memory: {available_memory / (1024 * 1024):,.2f} MB ({available_memory / (1024 * 1024 * 1024):.2f} GB)."
+        )
+
+        with open(local_file_path, "wb") as f:
+            # Measure the time for reading
+            logger.info("Starting to read the file")
+            start_read_time = time.time()
+            file_content = fp.read()  # Read the file content
```

Comment:
Yes, it makes sense to me; right now, the testing is limited to < 1GB per file and files total sum of < 5GB, but once we move to bigger files, we will need to change this. Let me check if I can do some quick tests for a change with reading chunks.

Reply:
> testing is limited to < 1GB per file and files total sum of < 5GB

AFAIK, Our source connectors start with 300Mb memory limit, so we likely reach OOM a lot sooner.

Reply:
I have added a [commit](https://github.com/airbytehq/airbyte/pull/47705/commits/e6de69bc5066c183f8bca0c7f2859461db475962) for this change on a different branch. That other branch is the one I use to create the dev versions of the connector to test in the cloud. If it works fine in the [testing workspace](https://cloud.airbyte.com/workspaces/166f3e04-9021-4f63-a963-a43abb548cec/connections), I will cherry-pick on this branch.

Reply:
I am concerned as this change seems to produce a problem with reading from the source significantly slowing down the performance: 
![image](https://github.com/user-attachments/assets/afb8b61a-c6a2-463e-9572-af8990bd31db)

vs usual speed with reading the full file:

![image](https://github.com/user-attachments/assets/2cefdd2f-d65e-45bb-b05c-666b0eac5ffb)

I still think we want to read/write by chunks. This seems to be a problem with the connector read library (paramiko), but I will need more time to investigate. For the first iteration requirements, I think we are good without the chunks reading.

The first iteration requires:

< 1GB per file
< 5GB total sum of files.

Reply:
> > testing is limited to < 1GB per file and files total sum of < 5GB
> 
> AFAIK, Our source connectors start with 300Mb memory limit, so we likely reach OOM a lot sooner.

Oh, I missed this comment, some changes have occurred in the connectors pods specifically for this project, I added some logging that can illustrate this point: 

`2024-10-30 16:54:23 source > Writing file to '/staging/files/5GB_Files//random_data_10_1.csv' with size: 950.73 MB (0.93 GB), available disk space: 26,790.53 MB (26.16 GB),available memory: 27,518.78 MB (26.87 GB).`


So for now I still think we are good.

Reply:
https://github.com/airbytehq/airbyte-internal-issues/issues/10480

Reply:
I think this makes sense, for my part. Short term, we can bump the memory on the connector, but long-term, writing by chunk and keeping memory footprint down is going to be necessary.

We may also want to consider letting subclasses (S3/SFTP, for instance) implement their own download process. Our generic implementation treats this as a generic file buffer (which lets us generically solve all use cases), but other implementations might have native approaches that are already tuned for optimum throughput / resiliency.

Reply:
I've updated the wording in the [follow-up item](https://github.com/airbytehq/airbyte-internal-issues/issues/10480).

Reply:
> available memory: 27,518.78 MB (26.87 GB).

😲 WOW😲 


I wish all our pods had this much memory. I still believe that a longer sync time is an acceptable trade-off, as my main concern is with the OSS and enterprise standalone versions: we cannot guarantee this level of memory for their deployments. Therefore, I’d prefer our connector to be stable and memory-efficient rather than fast.

But I don't want to block you, so feel free to leave a `TODO` and add the task to the epic. We can definitely come back to it later.

Reply:
added the todo and [the task](https://github.com/airbytehq/airbyte-internal-issues/issues/10480) was updated with the latest comments